src/main/java/org/openrewrite/staticanalysis/SpecifyGenericExceptionCatches.java

```diff
@@ -0,0 +1,199 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://docs.moderne.io/licensing/moderne-source-available-license
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.staticanalysis;
+
+import org.openrewrite.Cursor;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.JavaVisitor;
+import org.openrewrite.java.ShortenFullyQualifiedTypeReferences;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.JavaType.FullyQualified;
+import org.openrewrite.java.tree.TypeUtils;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+
+public class SpecifyGenericExceptionCatches extends Recipe {
+
+    @Override
+    public String getDisplayName() {
+        return "Replace `catch(Exception)` with specific exceptions thrown in the try block";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Replaces `catch(Exception e)` blocks with a multi-catch block " +
+                "(`catch (SpecificException1 | SpecificException2 e)`) containing only the checked exceptions " +
+                "explicitly thrown by method or constructor invocations within the `try` block that are not " +
+                "already caught by more specific `catch` clauses. If no checked exceptions are found that " +
+                "require catching, the generic `catch` block is removed.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return Collections.unmodifiableSet(new HashSet<>(Arrays.asList("CWE-396", "RSPEC-S2221")));
+    }
+
+    @Override
+    public JavaVisitor<ExecutionContext> getVisitor() {
+        return new GenericExceptionCatchVisitor();
+    }
+
+    private static class GenericExceptionCatchVisitor extends JavaIsoVisitor<ExecutionContext> {
+
+        private static final String JAVA_LANG_EXCEPTION = "java.lang.Exception";
+        private static final String MULTI_CATCH_SEPARATOR = "|";
+        private static final String TRY_CATCH_TEMPLATE = "try {} catch (%s %s) {}";
+
+        @Override
+        public J.Try visitTry(J.Try aTry, ExecutionContext ctx) {
+            J.Try t = super.visitTry(aTry, ctx);
+
+            if (hasGenericCatch(t)) {
+                Set<JavaType> caughtExceptions = getCaughtExceptions(t);
+                Set<JavaType> thrownExceptions = getThrownExceptions(t);
+                thrownExceptions.removeAll(caughtExceptions); // Remove exceptions that are already specifically caught
```

Comment:
Forgot to add maybeAddImport for these exceptions

Reply:
Although I guess it's not really needed as `ShortenFullyQualifiedTypeReferences` does it already